### DISCUSSION TITLE
Support for MongoDB 2.6 and newer

### DISCUSSION THREAD

**Discussion by xet7** [meteor/meteor#12353] [Feature Requests]
From @xet7

If I would like to add back supporting also older versions of MongoDB, like that Sandstorm https://sandstorm.io https://github.com/sandstorm-io/sandstorm is still at MongoDB 3.0 , what would be most practical way to add support back? WeKan Open Source kanban https://wekan.github.io https://github.com/wekan/wekan currenty is at newest Meteor 2.8.2 using Node.js 14.x and MongoDB 5.x.

From @zenhack

Sandstorm is currently stuck on 2.6, not 3.0, though meteor-spk uses 3.0

From @xet7

It's just that migrations could use a lot of limited disk space, for example in some case they have 500 GB size of MongoDB database, and converting that to newer MongoDB would take at least 4 hours and there is not enough disk space at same partition to do that migration. It would be useful to still support all MongoDB versions possible.

To try to make something, I have tried some different ways:

a) Alough it does not fully work yet, in [my Meteor fork](https://github.com/xet7/meteor) I made [some changes](https://github.com/xet7/meteor/commit/48aa5eb87edf0cc6ff9b711e906d40e3489f076f) to revert https://github.com/meteor/meteor/pull/11556 , remove limit for some CPU platforms, change various code from `createIndex` to `_ensureIndex`, and change jquery version to be more compatible with current WeKan version, etc. At [WeKan Mongo 3.0 Sandstorm branch](https://github.com/wekan/wekan/tree/wekan-mongo-3.0-sandstorm) I also made some similar changes for `ensureIndex` etc. But I don't know is this good idea.

b) I have been thinking, should I fork https://www.npmjs.com/package/mongodb , add back those deprecated MongoDB versions support, and add compatibility code so that both `createIndex` and `_ensureIndex` code would work the same way, it would check from MongoDB server version about what MongoDB wire protocol to use in that case.

From @zenhack

c) FerretDB https://github.com/meteor/meteor/discussions/12150 is a more promising replacement than lungo at this point; I think I dug into it and there was actually kindof a lot to build out for lungo

From @xet7

d) Should I try to add MongoDB wire protocol etc to Go based MongoDB alternative https://github.com/256dpi/lungo/issues/10 , so that Go database server would be replacement of MongoDB.

e) For doing MongoDB migration to newer one, there are a little code at https://github.com/sandstorm-io/sandstorm/pull/3651 but it would dump and restore database

f) I have made some script that would export from MongoDB to SQLite https://github.com/wekan/hx/blob/main/prototypes/database/MongoDB/mongoexport-csv-sqlite/mongoexport.sh and have some very minimal PHP prototype using that SQLite database https://github.com/wekan/php/blob/main/page/allboardschar.php . Well, kind of because some would like to have WeKan version without websockets, because those are forbidden in some internal network, and there has been a lot of problems with loading too much data (all cards of a big board) too often with subscriptions, slowing down network with high bandwidth usage. Also there has been some problems uploading files with websockets related CollectionFS or Meteor-Files, sometime that code has automatically changed from websockets to long-polling, or there has been incomplete uploads that would need to be removed before wekan would start properly at all.

g) I have been thinking would it be possible to add SQLite support to Meteor, that it would work better at Raspberry Pi and similar arm64 boards.

h) Then I also have minimal Meteor example https://github.com/wekan/we that is based on newest Meteor, but does not have any MongoDB packages, other than older version of MongoDB driver, that can make database queries at Sandstorm. Although, that then does not have remaining of WeKan, that remaining would require remaining newest Meteor MongoDB related packages.

Sure it seems that Meteor is currently going to direction of newest Node.js and MongoDB etc. Just thinking does anyone have some helpful tips about what would be most practical way do something about this.

Well, I'm also in progress of trying to optimize WeKan for bigger amount of users, like 12k or 30k, move code from clientside to serverside, etc.

**Comment by radekmie** [comment]
I fully get where is this coming from, but... MongoDB 2.6 is not even listed on the [Lifecycle Schedules](https://www.mongodb.com/support-policy/lifecycles) and 3.0 EOLd nearly 5 years ago. Maybe it'd be better to put this effort into making it work with newer MongoDB versions instead of making everything else work with older MongoDB version?

**Reply by ocdtrekkie** [reply]
So, part of the problem is that current MongoDB offerings are SSPL licensed, and Meteor dropped support for all of them at the same time, making it impossible to run fully open source software with Meteor. Is there a newer open source version or database Meteor could support?

**Reply by radekmie** [reply]
Is SSPL truly a problem for you? I know it's not as permissive as MIT, but is this affecting you in any way?

Nevertheless, you may look into FerretDB: https://github.com/meteor/meteor/discussions/12150. There's even a [Slack channel](https://meteor-community.slack.com/archives/C04B847PQRL) dedicated for it and people behind it are really responsive.

**Reply by zenhack** [reply]
I think longer term we'd like to move to something like FerretDB, but it's not ready yet; I'd hoped we'd be able to buy a bit of time by upgrading to 4.0  or such (which iirc was the last agpl3 version) but meteor dropped support for 3 major releases at once...

TBH, I'd probably balk at this suggestion too; I wouldn't be keen on supporting "so old it's EOL date has been forgotten" either.

Even beyond the license issues, upgrading is a pain because we need to fully automate it, and it has to be reliable enough to have negligible chance of failing; we're bundling mongo and shipping it to users who may not have sysadmin skills. It might honestly be less work for us to maintain a meteor fork that kept this support until we can cut over to something else entirely than to keep fussing with mongo upgrades until then (it also might not; I'm not familiar enough with the code to really gauge).

**Reply by zenhack** [reply]
Re: licensing, the problem is mostly that it repels some of our otherwise most likely users just by being "not actually FOSS."

**Reply by ocdtrekkie** [reply]
> Is SSPL truly a problem for you? I know it's not as permissive as MIT, but is this affecting you in any way?

So I don't mind (and even like) the SSPL *personally*, and it doesn't impact our use case in practice, but it can be very challenging to maintain and build open source communities around licenses the OSI has decided to reject.

I'd say that while the goals of SSPL are noble, it practically inhibits open source communities, and that's problematic for us *as a project*.

**Reply by radekmie** [reply]
OK, got it now. Maybe it'd be easier to do something like a compatibility layer but outside of the app? Like, FerretDB behaves like MongoDB but translates it to Postgres commands. You could do the same to translate new commands down to the old format...? I don't know which of the options would be best and safest, really.

If it was me, I'd be much more concerned about support and security updates than being truly open source. But I fully understand both you and your clients in this matter.
